from sqlalchemy import create_engine, text

database_name = 'scooters'
connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

connection_string

engine = create_engine(connection_string)

engine

#Initial EDA
query = '''
SELECT *
FROM scooters
LIMIT 10;
'''

with engine.connect() as connection:
    result = connection.execute(text(query))

result.fetchall()

In [1]:
import pandas as pd

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Bird';
'''

with engine.connect() as connection:
    bird = pd.read_sql(text(query), con = connection)

bird.head()

bird.to_csv("birddf.csv", index = False)

In [2]:
#read in bird
birddf = pd.read_csv('../data/birddf.csv')

In [3]:
birddf.shape

(12251590, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Bolt';
'''

with engine.connect() as connection:
    bolt = pd.read_sql(text(query), con = connection)

bolt.head()

bolt.to_csv("boltdf.csv", index = False)

In [4]:
#Read in bolt
boltdf = pd.read_csv('../data/boltdf.csv')

In [5]:
boltdf.shape

(3477198, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Gotcha';
'''

with engine.connect() as connection:
    gotcha = pd.read_sql(text(query), con = connection)

gotcha.head()

gotcha.to_csv("gotchadf.csv", index = False)

In [6]:
#Read in gotcha
gotchadf = pd.read_csv('../data/gotchadf.csv')

In [7]:
gotchadf.shape

(4679280, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Jump';
'''

with engine.connect() as connection:
    jump = pd.read_sql(text(query), con = connection)

jump.head()

#Saving jump to csv

jump.to_csv("jumpdf.csv", index = False)

In [8]:
#Read in jump
jumpdf = pd.read_csv('../data/jumpdf.csv')

In [9]:
jumpdf.shape

(21835098, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Lime';
'''

with engine.connect() as connection:
    lime = pd.read_sql(text(query), con = connection)

lime.head()

lime.to_csv("limedf.csv", index = False)

In [10]:
limedf = pd.read_csv('../data/limedf.csv')

In [11]:
limedf.shape

(16524261, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Lyft';
'''

with engine.connect() as connection:
    lyft = pd.read_sql(text(query), con = connection)

lyft.head()

lyft.to_csv("lyftdf.csv", index = False)

In [12]:
lyftdf = pd.read_csv('../data/lyftdf.csv')

In [13]:
lyftdf.shape

(9087043, 9)

query = '''
SELECT *
FROM scooters
WHERE companyname = 'Spin';
'''

with engine.connect() as connection:
    spin = pd.read_sql(text(query), con = connection)

spin.head()

spin.to_csv("spindf.csv", index = False)

In [14]:
spindf = pd.read_csv('../data/spindf.csv')

In [15]:
spindf.shape

(5559573, 9)

query = '''
SELECT *
FROM trips;
'''

with engine.connect() as connection:
    tripsdf = pd.read_sql(text(query), con = connection)

tripsdf.head()

tripsdf.to_csv("tripsdf.csv", index = False)

query = '''
SELECT *
FROM trips;
'''

with engine.connect() as connection:
    trips = pd.read_sql(text(query), con = connection)

trips.head()

In [16]:
tripsdf = pd.read_csv('../data/tripsdf.csv')

In [17]:
tripsdf.shape

(565522, 16)

In [18]:
tripsdf.info

<bound method DataFrame.info of                    pubtimestamp companyname triprecordnum  \
0       2019-05-01 00:00:55.423        Bird       BRD2134   
1       2019-05-01 00:03:33.147        Lyft          LFT5   
2       2019-05-01 00:05:55.570        Bird       BRD2168   
3       2019-05-01 00:05:55.570        Bird       BRD2166   
4       2019-05-01 00:05:55.570        Bird       BRD2165   
...                         ...         ...           ...   
565517  2019-08-01 04:53:48.000        JUMP          JMP3   
565518  2019-08-01 04:53:48.000        JUMP          JMP2   
565519  2019-08-01 04:53:48.000        JUMP          JMP1   
565520  2019-08-01 07:04:00.000        JUMP         JMP35   
565521  2019-08-01 07:04:00.000        JUMP         JMP34   

                                             sumdid  tripduration  \
0                                      Powered9EAJL      3.000000   
1                                     Powered296631      1.715600   
2                           

In [19]:
#Concat to create the scooters table
scooters_list = [birddf, boltdf, gotchadf, jumpdf, limedf, lyftdf, spindf]
scootersdf = pd.concat(scooters_list)

In [20]:
scootersdf.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
0  2019-07-30 17:37:28.870  36.141562 -86.806655  PoweredVAW9Y  Powered   
1  2019-07-30 17:37:28.870  36.159744 -86.780067  PoweredSNB79  Powered   
2  2019-07-30 17:37:28.870  36.153569 -86.784961  PoweredZY5KX  Powered   
3  2019-07-30 17:37:28.870  36.155984 -86.782756  PoweredPHVAG  Powered   
4  2019-07-30 17:37:28.870  36.147253 -86.815000  Powered238ZE  Powered   

   chargelevel sumdgroup  costpermin companyname  
0         95.0   scooter        0.15        Bird  
1         58.0   scooter        0.15        Bird  
2        100.0   scooter        0.15        Bird  
3        100.0   scooter        0.15        Bird  
4         76.0   scooter        0.15        Bird

In [21]:
scootersdf.info()

<class 'pandas.core.frame.DataFrame'>
Index: 73414043 entries, 0 to 5559572
Data columns (total 9 columns):
 #   Column       Dtype  
---  ------       -----  
 0   pubdatetime  object 
 1   latitude     float64
 2   longitude    float64
 3   sumdid       object 
 4   sumdtype     object 
 5   chargelevel  float64
 6   sumdgroup    object 
 7   costpermin   float64
 8   companyname  object 
dtypes: float64(4), object(5)
memory usage: 5.5+ GB


## - Are there any null values in any columns in either table?

In [22]:
#Checking for NaN values in the scooters dataframe
scootersdf.isna().any()

pubdatetime    False
latitude       False
longitude      False
sumdid         False
sumdtype       False
chargelevel     True
sumdgroup      False
costpermin     False
companyname    False
dtype: bool

In [23]:
scootersdf.isnull().sum().sum()

770

In [24]:
#Checking for NaN values in the trips dataframe
tripsdf.isna().any()

pubtimestamp      False
companyname       False
triprecordnum     False
sumdid            False
tripduration      False
tripdistance      False
startdate         False
starttime         False
enddate           False
endtime           False
startlatitude     False
startlongitude    False
endlatitude       False
endlongitude      False
triproute         False
create_dt         False
dtype: bool

## - What date range is represented in each of the date columns? Investigate any values that seem odd.

In [25]:
scootersdf['pubdatetime'] = pd.to_datetime(scootersdf['pubdatetime'], dayfirst = False, format = 'mixed')

In [26]:
scootersdf.dtypes

pubdatetime    datetime64[ns]
latitude              float64
longitude             float64
sumdid                 object
sumdtype               object
chargelevel           float64
sumdgroup              object
costpermin            float64
companyname            object
dtype: object

In [27]:
scootersdf['pubdatetime'].max()

Timestamp('2019-07-31 23:59:57')

In [28]:
scootersdf['pubdatetime'].min()

Timestamp('2019-05-01 00:01:41.247000')

In [29]:
tripsdf['pubtimestamp'] = pd.to_datetime(tripsdf['pubtimestamp'], dayfirst = False, format = 'mixed')

In [30]:
tripsdf.dtypes

pubtimestamp      datetime64[ns]
companyname               object
triprecordnum             object
sumdid                    object
tripduration             float64
tripdistance             float64
startdate                 object
starttime                 object
enddate                   object
endtime                   object
startlatitude            float64
startlongitude           float64
endlatitude              float64
endlongitude             float64
triproute                 object
create_dt                 object
dtype: object

#Checking out the pubtimestamp column of the trips table
tripsdf["pubtimestamp"]

#Checking out the create_dt column in the trips df
tripsdf["create_dt"]

In [31]:
tripsdf['pubtimestamp'].min()

Timestamp('2019-05-01 00:00:55.423000')

In [32]:
tripsdf['pubtimestamp'].max()

Timestamp('2019-08-01 07:04:00')

The scooters df date range is '2019-05-01 00:01:41.247000' - '2019-07-31 23:59:57'.
The trips df date range is '2019-05-01 00:00:55.423000' - '2019-08-01 07:04:00'

## - Is time represented with am/pm or using 24 hour values in each of the columns that include time?

Time is represented in 24 hour values in the time columns in both dataframes.

## - What values are there in the sumdgroup column? Are there any that are not of interest for this project?

In [33]:
scootersdf.sumdgroup.unique()

array(['scooter', 'Scooter', 'bicycle'], dtype=object)

We don't need the bicycle sumdgroup for this project.

## - What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [34]:
scootersdf['latitude'].min()

0.0

In [35]:
scootersdf['latitude'].max()

3609874.116666

In [36]:
scootersdf['longitude'].min()

-97.443879

In [37]:
scootersdf['longitude'].max()

0.0

In [38]:
tripsdf['startlatitude'].min()

35.8532

In [39]:
tripsdf['startlatitude'].max()

36.300029

In [40]:
tripsdf['startlongitude'].min()

-86.918008

In [41]:
tripsdf['startlongitude'].max()

-86.3662

The minimum latitude shown is "0.0", as well as the maximum longitude, which is surprising. The maximum latitude is '3609874.116666' which is also "surprising" to say the least.

## -What is the range of values for trip duration and trip distance? Do these values make sense? Explore values that might seem questionable.

In [42]:
tripsdf['tripduration'].min()

-19.3582666667

In [43]:
tripsdf['tripduration'].max()

512619.0

In [44]:
tripsdf['tripdistance'].min()

-20324803.8

In [45]:
tripsdf['tripdistance'].max()

31884482.6476

In [46]:
tripsdf.query('`tripduration`== -19.3582666667')

pubtimestamp companyname triprecordnum         sumdid  \
509599 2019-07-19 00:14:02.297        Lyft         LFT21  Powered853770   

        tripduration  tripdistance   startdate        starttime     enddate  \
509599    -19.358267    4540.68256  2019-07-19  00:09:04.506666  2019-07-18   

                endtime  startlatitude  startlongitude  endlatitude  \
509599  23:49:43.013333       36.15505       -86.78482      36.1469   

        endlongitude                                          triproute  \
509599     -86.79355  [(36.15505, -86.78482), (36.15511, -86.78482),...   

                      create_dt  
509599  2019-07-20 10:52:39.737

In [47]:
tripsdf.query('`tripdistance`== -20324803.8')

pubtimestamp companyname triprecordnum        sumdid  \
550038 2019-07-27 23:56:41.940        Bird       BRD5324  PoweredYZL2A   

        tripduration  tripdistance   startdate        starttime     enddate  \
550038          44.0   -20324803.8  2019-07-27  23:14:23.683333  2019-07-27   

                endtime  startlatitude  startlongitude  endlatitude  \
550038  23:58:23.820000        36.1585        -86.7683       36.126   

        endlongitude                                          triproute  \
550038      -86.7892  [(36.158517, -86.768224), (36.158624, -86.7682...   

                      create_dt  
550038  2019-07-28 05:32:05.610

In [48]:
tripsdf.query('`tripdistance`== 31884482.6476')

pubtimestamp companyname triprecordnum       sumdid  \
447799 2019-07-05 01:18:52      Gotcha          GOT9  Powered2460   

        tripduration  tripdistance   startdate starttime     enddate  \
447799           9.0  3.188448e+07  2019-07-05  01:18:56  2019-07-05   

         endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
447799  01:27:56       36.16338      -86.780142     36.16628    -86.778127   

                                      triproute                create_dt  
447799  [('0', '0'), ('36.16628', '-86.77813')]  2019-07-06 07:13:12.210

## - Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?


In [49]:
scootersdf.companyname.unique()

array(['Bird', 'Bolt', 'Gotcha', 'Jump', 'Lime', 'Lyft', 'Spin'],
      dtype=object)

In [50]:
tripsdf.companyname.unique()

array(['Bird', 'Lyft', 'Lime', 'Bolt Mobility', 'SPIN', 'Gotcha', 'JUMP'],
      dtype=object)

In the trips df, "Bolt" is "Bolt Mobility", "Jump", and "Spin" are also in upper case.

## 1. During this period, seven companies offered scooters. How many scooters did each company have in this time frame? Did the number for each company change over time? Did scooter usage vary by company?

In [51]:
scootersdf.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
0 2019-07-30 17:37:28.870  36.141562 -86.806655  PoweredVAW9Y  Powered   
1 2019-07-30 17:37:28.870  36.159744 -86.780067  PoweredSNB79  Powered   
2 2019-07-30 17:37:28.870  36.153569 -86.784961  PoweredZY5KX  Powered   
3 2019-07-30 17:37:28.870  36.155984 -86.782756  PoweredPHVAG  Powered   
4 2019-07-30 17:37:28.870  36.147253 -86.815000  Powered238ZE  Powered   

   chargelevel sumdgroup  costpermin companyname  
0         95.0   scooter        0.15        Bird  
1         58.0   scooter        0.15        Bird  
2        100.0   scooter        0.15        Bird  
3        100.0   scooter        0.15        Bird  
4         76.0   scooter        0.15        Bird

In [52]:
scootersdf['Month'] = scootersdf['pubdatetime'].dt.month

In [53]:
scooters_may = scootersdf[scootersdf['Month'] == 5]

In [54]:
scooters_may.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
145127 2019-05-01 00:01:41.247  36.136822 -86.799877  PoweredLIRL1  Powered   
145128 2019-05-01 00:01:41.247  36.191252 -86.772945  PoweredXWRWC  Powered   
145129 2019-05-01 00:01:41.247  36.144752 -86.806293  PoweredMEJEH  Powered   
145130 2019-05-01 00:01:41.247  36.162056 -86.774688  Powered1A7TC  Powered   
145131 2019-05-01 00:01:41.247  36.150973 -86.783109  Powered2TYEF  Powered   

        chargelevel sumdgroup  costpermin companyname  Month  
145127         93.0   scooter         0.0        Bird      5  
145128         35.0   scooter         0.0        Bird      5  
145129         90.0   scooter         0.0        Bird      5  
145130         88.0   scooter         0.0        Bird      5  
145131         98.0   scooter         0.0        Bird      5

In [55]:
may_company_scooters = scooters_may[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'maycount')\
                                                        .sort_values(['maycount'])

In [56]:
may_company_scooters

companyname  maycount
2      Gotcha       223
1        Bolt       346
6        Spin       687
3        Jump       976
5        Lyft      1376
4        Lime      1632
0        Bird      3064

In [57]:
scooters_june = scootersdf[scootersdf['Month'] == 6]

In [58]:
june_company_scooters = scooters_june[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'junecount')\
                                                        .sort_values(['junecount'])

In [59]:
june_company_scooters

companyname  junecount
2      Gotcha        223
1        Bolt        333
6        Spin        713
3        Jump       1036
5        Lyft       1152
4        Lime       1461
0        Bird       2910

In [60]:
scooters_july = scootersdf[scootersdf['Month'] == 7]

In [61]:
july_company_scooters = scooters_july[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'julycount')\
                                                        .sort_values(['julycount'])

In [62]:
july_company_scooters

companyname  julycount
2      Gotcha        224
1        Bolt        276
6        Spin        570
4        Lime        903
5        Lyft       1026
3        Jump       1206
0        Bird       2583

In [63]:
may_june = may_company_scooters.merge(june_company_scooters, how = 'inner', on = 'companyname')

In [64]:
company_scooters_months = may_june.merge(july_company_scooters, how = 'inner', on = 'companyname')

In [65]:
company_scooters_months

companyname  maycount  junecount  julycount
0      Gotcha       223        223        224
1        Bolt       346        333        276
2        Spin       687        713        570
3        Jump       976       1036       1206
4        Lyft      1376       1152       1026
5        Lime      1632       1461        903
6        Bird      3064       2910       2583

In [66]:
tripsdf

pubtimestamp companyname triprecordnum  \
0      2019-05-01 00:00:55.423        Bird       BRD2134   
1      2019-05-01 00:03:33.147        Lyft          LFT5   
2      2019-05-01 00:05:55.570        Bird       BRD2168   
3      2019-05-01 00:05:55.570        Bird       BRD2166   
4      2019-05-01 00:05:55.570        Bird       BRD2165   
...                        ...         ...           ...   
565517 2019-08-01 04:53:48.000        JUMP          JMP3   
565518 2019-08-01 04:53:48.000        JUMP          JMP2   
565519 2019-08-01 04:53:48.000        JUMP          JMP1   
565520 2019-08-01 07:04:00.000        JUMP         JMP35   
565521 2019-08-01 07:04:00.000        JUMP         JMP34   

                                             sumdid  tripduration  \
0                                      Powered9EAJL      3.000000   
1                                     Powered296631      1.715600   
2                                      Powered7S2UU      3.000000   
3                                      PoweredZIIVX      3.000000   
4                                      PoweredJ7MB3      2.000000   
...                                             ...           ...   
565517  Powered5614e4fc-1971-5e5c-a545-0fd88cd3331d     17.497950   
565518  Powereda4712099-5e55-5332-996c-d6e5c910535d      3.154917   
565519  Poweredb8a3a269-d1ca-571f-9b2f-89b7399b5537     18.440650   
565520  Poweredf077a919-d569-5e70-8ca7-71d179ffacf9    142.345610   
565521  Powered784fcc92-bd1a-5f6a-8314-b02aa21b4bfa    209.409650   

        tripdistance   startdate        starttime     enddate  \
0          958.00528  2019-05-01  00:00:20.460000  2019-05-01   
1         1371.39112  2019-05-01  00:01:50.090000  2019-05-01   
2         2296.58800  2019-05-01  00:03:47.363333  2019-05-01   
3         1200.78744  2019-05-01  00:04:21.386666  2019-05-01   
4          351.04988  2019-05-01  00:04:27.796666  2019-05-01   
...              ...         ...              ...         ...   
565517    7075.20000  2019-07-31         23:47:43  2019-08-01   
565518    1320.00000  2019-07-31         23:57:19  2019-08-01   
565519    7920.00000  2019-07-31         23:41:52  2019-08-01   
565520   20433.60000  2019-07-31         23:26:15  2019-08-01   
565521   14889.60000  2019-07-31         22:19:08  2019-08-01   

                endtime  startlatitude  startlongitude  endlatitude  \
0       00:02:52.346666      36.157100      -86.803600    36.156600   
1       00:03:33.026666      36.157970      -86.778960    36.160540   
2       00:07:13.596666      36.154700      -86.781800    36.156500   
3       00:06:59.176666      36.149400      -86.779500    36.153100   
4       00:06:23.150000      36.177800      -86.786600    36.177400   
...                 ...            ...             ...          ...   
565517         00:05:13      36.154633      -86.798340    36.149220   
565518         00:00:28      36.153015      -86.783585    36.152843   
565519         00:00:18      36.153687      -86.784580    36.164158   
565520         01:48:35      36.155735      -86.775185    36.173904   
565521         01:48:33      36.160050      -86.776720    36.163410   

        endlongitude                                          triproute  \
0         -86.806700  [(36.157235, -86.803612), (36.157235, -86.8036...   
1         -86.776890  [(36.15797, -86.77896), (36.15795, -86.77873),...   
2         -86.786800  [(36.155068, -86.782124), (36.156597, -86.78675)]   
3         -86.779600  [(36.149741, -86.779344), (36.149741, -86.7793...   
4         -86.787600  [(36.177699, -86.786477), (36.177711, -86.7864...   
...              ...                                                ...   
565517    -86.813980  [('36.154631', '-86.798341'), ('36.154631', '-...   
565518    -86.779594  [('36.153016', '-86.783586'), ('36.153016', '-...   
565519    -86.776900  [('36.153685', '-86.784578'), ('36.153685', '-...   
565520    -86.785450  [('36.155736', '-86.775181'), ('36.155531', '-...   
565521   

In [102]:
#Counting pubtimestap per company
tripsdf[['companyname', 'pubtimestamp']].groupby(['companyname'])['pubtimestamp']\
                                                        .nunique()

companyname
Bird              17242
Bolt Mobility      8072
Gotcha             2489
JUMP                239
Lime              68852
Lyft             120925
SPIN               9993
Name: pubtimestamp, dtype: int64

## 2. According to Second Substitute Bill BL2018-1202 (as amended) (https://web.archive.org/web/20181019234657/https://www.nashville.gov/Metro-Clerk/Legislative/Ordinances/Details/7d2cf076-b12c-4645-a118-b530577c5ee8/2015-2019/BL2018-1202.aspx), all permitted operators will first clean data before providing or reporting data to Metro. Data processing and cleaning shall include:  
* Removal of staff servicing and test trips  
* Removal of trips below one minute  
* Trip lengths are capped at 24 hours  
Are the scooter companies in compliance with the second and third part of this rule?

In [67]:
#Looking for trips below one minute.
short_trips = tripsdf.query('`tripduration` < 1')

In [68]:
short_trips[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'count')\
                                                        .sort_values(['count'])

companyname  count
1        Lime    383
0        Bird   1229
2        Lyft   1329

In [69]:
#Looking for trips longer than 24 hours
long_trips = tripsdf.query('`tripduration` > 1440')

In [70]:
long_trips[['companyname', 'sumdid']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'count')\
                                                        .sort_values(['count'])

companyname  count
1           Lyft      2
2           SPIN     28
0  Bolt Mobility    351

## Answer: Companies (Bird, Bolt, Lime, Lyft, and Spin) failed to remove both trips shorter than 1 minute (Bird, Lime, & Lyft), and longer than 24 hours (Bolt, Lyft, & Spin) from this dataset.

## 3. The goal of Metro Nashville is to have each scooter used a minimum of 3 times per day. Based on the data, what is the average number of trips per scooter per day? Make sure to consider the days that a scooter was available. How does this vary by company?


From Postgres: Bird has 152,745 sumdids (3,766 distinct)
               Bolt has 21,890 sumdids (356 distinct)
               Gotcha has 3,315 sumdids (166 distinct)
               Jump has 6,437 sumdids (450 distinct)
               Lime has 225,694 sumdids (1,788 distinct)
               Lyft has 120,991 sumdids (1,725 distinct)
               Spin has 34,450 sumdids (754 distinct)

In [98]:
tripsdf.groupby('sumdid').size()

sumdid
Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619    92
Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d    57
Powered-03be23ca-d43b-222f-be54-e44b5b4690df    81
Powered-046201fb-6532-1f37-6334-3612fb1e61f7    71
Powered-0479bb84-afbd-0426-f1c4-df628542a88c    96
                                                ..
Poweredff356f94-2252-5432-b484-ad9679bf4b8d     11
Poweredff4b01f3-1f3f-5d6b-b1c5-5d6628ffa629      6
Poweredff8a28ab-0ffb-59ce-9b1e-2d1299a25634     23
Poweredffcc1b11-af95-5e06-b002-82ffb342aae8      6
StandardLIRPOZ7X3ZVGO                            1
Length: 9005, dtype: int64

tripsdf.pivot(index='sumdid', columns='companyname', values='pubtimestamp')

In [92]:
tripsdf["sumdid"].value_counts()

sumdid
PoweredUNKNOWN                                 3613
Poweredb671ffe4-2bca-5880-98ec-d36f65aa11a2    1908
PoweredR3QXO6TL2HIAB                            382
PoweredZ6HF5P5HCJCT5                            369
PoweredCL37MXJKM7RS5                            352
                                               ... 
PoweredIDYLX                                      1
PoweredS3KN7NPZZXTZ2                              1
Powered435266                                     1
Powered44GAM                                      1
PoweredQR6KS                                      1
Name: count, Length: 9005, dtype: int64

trips_days = (tripsdf
 .assign(day = tripsdf['pubtimestamp'].dt.day_name())
 .day
 .value_counts(sort = False)
)

trips_days

(tripsdf
 .groupby(pd.Grouper(key = 'pubtimestamp',     
                     freq = '1d',               
                     origin = 'epoch'          
                                               
                    ))
 ['sumdid']
 .count()
 .reset_index()
 .head(10)
)

(tripsdf
 .assign(day_name = tripsdf['pubtimestamp'].dt.day_name())
 .groupby('day_name')
 ['sumdid']
 .nunique()
)

tripsdf[['companyname', 'sumdid', 'pubtimestamp']].groupby(['companyname'])['sumdid']\
                                                        .nunique()\
                                                        .reset_index(name = 'count')\
                                                        .sort_values(['count'])

## 4. Metro would like to know how many scooters are needed, and something that could help with this is knowing peak demand. Estimate the highest count of scooters being used at the same time. When were the highest volume times? Does this vary by zip code or other geographic region?

## 5. **Stretch Goal:** SUMDs can provide alternative transportation and provide "last mile" access to public transit. How often are trips starting near public transit hubs? You can download a dataset of bus stop locations from https://data.nashville.gov/Transportation/WeGo-Transit-Bus-Stops/vfe9-k7vc/about_data.

## Deliverables:
At the conclusion of this project, your group should deliver a presentation which addresses the following points:
* Are scooter companies in compliance with the required data cleaning?
* What are typical usage patterns for scooters in terms of time, location, and trip duration?
* What are your recommendations for total number of scooters for the city overall and density of scooters by zip code?
* **Stretch Goal:** Does it appear that scooters are used as "last mile" transportation from public transit hubs to work or school?


-Only Gotcha and Jump are in compliance with the required data cleaning.

-

-

-